### ISSUE TITLE
Non-deterministic results when hashing with seed multiple categorical columns

### ISSUE THREAD

**Issue by gautiervarjo** [pola-rs/polars#7758]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

I'm hashing (with a seed) several categorical columns concurrently, and I was surprised to see that the hash changes between sessions.

I managed to reproduce the issue on a toy dataframe and observed that the string cache is part of the problem; within one session it's possible to get different hashes for a given (logical) value.

The reproduction example is of course a bit contrived, but it's not far from my actual use-case! :slightly_smiling_face: 

### Reproducible example

```python
import polars as pl

df = pl.DataFrame(data={"a": ["a"], "b": ["b"]})

for i in range(10):
    with pl.StringCache():
        result = (
            df
            .with_columns(pl.all().cast(pl.Categorical))
            .with_columns(pl.all().hash(seed=1).suffix("_hash"))
        )
        print(result["a_hash"][0], result["b_hash"][0])

# Sample output from my machine:
# 2437957984917655262 0
# 0 2437957984917655262
# 0 2437957984917655262
# 0 2437957984917655262
# 2437957984917655262 0
# 0 2437957984917655262
# 0 2437957984917655262
# 2437957984917655262 0
# 0 2437957984917655262
# 2437957984917655262 0
```

### Expected behavior

- I expected the hash to be computed on the logical values ("a" and "b"), not on the physical category integer.
- I expected the hash to be stable within a session.
- I expected the hash to be stable across consecutive sessions (same machine, no version changes, etc), but based on https://github.com/pola-rs/polars/issues/5345#issuecomment-1293412513 I understand if polars can't really promise that.

### Installed versions

<details>

```
---Version info---
Polars: 0.16.14
Index type: UInt32
Platform: Linux-5.15.0-67-generic-x86_64-with-glibc2.29
Python: 3.8.10 (default, Mar 13 2023, 10:26:41) 
[GCC 9.4.0]
---Optional dependencies---
numpy: 1.22.3
pandas: 1.4.2
pyarrow: 11.0.0
connectorx: <not installed>
deltalake: <not installed>
fsspec: 2022.10.0
matplotlib: 3.5.1
xlsx2csv: <not installed>
xlsxwriter: <not installed>
```

</details>

**Comment by alexander-beedie** [comment]
Well-found; as that comment says, we do not currently offer deterministic hashing across architectures/sessions/versions... it's intra-session stable only for now, prioritising raw speed. We should make that clearer in the docs 🤔

**Comment by ion-elgreco** [comment]
> Well-found; as that comment says, we do not currently offer deterministic hashing across architectures/versions... it's intra-session stable only for now, prioritising raw speed. We should make that clearer in the docs 🤔

So if I hash the identical records across different versions, I will never get the same results?

**Comment by avimallu** [comment]
I think it means that it's not guaranteed, not that it will never happen.

**Comment by ion-elgreco** [comment]
> I think it means that it's not guaranteed, not that it will never happen. 

I see, that's still slightly problematic for the use case I was using hashing for. I think I'll need to replace it with an `.apply` again.

**Comment by ritchie46** [comment]
Not checked, but I think we might have stable hash between sessions if the seed is set. (I think we should).

However, it can be that your categorical levels will differ because of variability in jobs fetched in parallel. If we convert multiple string columns to categoricals in parallel. The order of columns will change as sometimes a thread will pick up column A sooner than B. This is the nature of parallelism and if you want stable levels, you must convert your categoricals in serial manner, i.e. one column per context.

**Comment by ion-elgreco** [comment]
> Not checked, but I think we might have stable hash between sessions if the seed is set. (I think we should).
> 
> However, it can be that your categorical levels will differ because of variability in jobs fetched in parallel. If we convert multiple string columns to categoricals in parallel. The order of columns will change as sometimes a thread will pick up column A sooner than B. This is the nature of parallelism and if you want stable levels, you must convert your categoricals in serial manner, i.e. one column per context.

For the non-categorical dtypes I did a sanity check by hashing the same dataframe across different versions and all of them were giving the same result : )

**Comment by orlp** [comment]
Closing as a duplicate of https://github.com/pola-rs/polars/issues/7758.
